### PR TITLE
feat: record layer preprocessing

### LINKED ISSUE #444: Record layer MPC preprocessing
Right now we already support preprocessing the protocols for the handshake, however we still perform all MPC processing for the record layer in the online phase. This can cause connection timeouts with the server.

Implement preprocessing for the record layer as well.

### PR DESCRIPTION
**Pull request by sinui0**
Closes #444 

This PR implements MPC preprocessing for the record layer. Specifically, it will pregarble the circuits for encryption up to the configured transcript limit and 256 bytes for decryption for the handshake.

Later we will want to come up with a strategy for preprocessing decryption, keeping in mind that a User may want to decrypt before and after the TLS connection is closed. This interacts with the deferred decryption feature in an annoying way.

### CODE REVIEW

**Review by th4s — APPROVED** [review]
Looking good, gw :+1:

**Comment by themighty1** [comment]
will review soon.

File: components/cipher/stream-cipher/src/stream_cipher.rs [hunk 1]
```diff
@@ -54,56 +53,50 @@ struct KeyAndIv {
     iv: Vec<u8>,
 }
 
+struct Transcript {
+    id: String,
+    plaintext: NestedId,
+}
```
Comment (themighty1) [inline components/cipher/stream-cipher/src/stream_cipher.rs:59]:
```suggestion
/// A subset of plaintext bytes processed by the stream cipher.
///
/// Note that `Transcript` does not store the actual bytes. Instead, it provides IDs which are
/// assigned to plaintext bytes of the stream cipher.
struct Transcript {
    /// The ID of this transcript.
    id: String,
    /// The ID for the next plaintext byte.
    plaintext: NestedId,
}
```

File: components/cipher/stream-cipher/src/stream_cipher.rs [hunk 2]
```diff
@@ -114,108 +107,141 @@ where
...
 
         Ok(keystream)
     }
 
-    /// Applies the keystream to the provided input text.
-    #[cfg_attr(
-        feature = "tracing",
-        tracing::instrument(level = "trace", skip(self), err)
-    )]
     async fn apply_keystream(
```
Comment (themighty1) [inline components/cipher/stream-cipher/src/stream_cipher.rs:132]:
```suggestion
    /// Applies the keystream to the provided input text.
    async fn apply_keystream(
```

**Review by themighty1 — CHANGES_REQUESTED** [review]
gw, some comments and maybe a code change.

File: components/cipher/stream-cipher/src/stream_cipher.rs [hunk 3]
```diff
@@ -114,108 +107,141 @@ where
...
 
     async fn verify(&mut self, value: ValueRef, expected: Value) -> Result<(), StreamCipherError> {
-        let mut scope = self.thread_pool.new_scope();
-        scope.push(move |thread| {
-            Box::pin(async move { thread.verify(&[value], &[expected]).await })
-        });
-        scope.wait().await.into_iter().next().unwrap()?;
+        self.thread_pool
+            .get_mut()
+            .verify(&[value], &[expected])
```
[the remark this answers is not in the document]
Comment (themighty1) [reply components/cipher/stream-cipher/src/stream_cipher.rs:256]: ah, ok, so the `Verify` trait guarantees both authenticity and correctness of the output. I thought it only dealt with authenticity as per its description.
Are you ok if i change the traits description to explicitely mention correctness @sinui0

File: components/cipher/stream-cipher/src/config.rs [hunk 4]
```diff
@@ -82,3 +46,10 @@ impl std::fmt::Debug for InputText {
         }
     }
 }
+
+#[derive(Debug, Clone, Copy)]
+pub(crate) enum ExecutionMode {
+    Mpc,
+    Prove,
+    Verify,
```
[the remark this answers is not in the document]
Comment (themighty1) [reply components/cipher/stream-cipher/src/config.rs:54]: >  Prove -> only this party has private inputs

That description does not fit here:

https://github.com/tlsnotary/tlsn/blob/8df19a04e5db7e56a742f5e6dff96660d80a783b/components/cipher/stream-cipher/src/keystream.rs#L254-L262

The `key` and `iv` are not private inputs of this party only, they are both parties' private inputs.

@sinui0

**Comment by themighty1** [comment]
gw, ACK to merge.
We can continue discussing a more precise type description in this PR.

**Review by themighty1 — APPROVED** [review]

**Pull request merged**